controller/cmd/public-api/main.go

```diff
@@ -26,6 +26,7 @@ func main() {
 	destinationAPIAddr := flag.String("destination-addr", "127.0.0.1:8086", "address of destination service")
 	tapAddr := flag.String("tap-addr", "127.0.0.1:8088", "address of tap service")
 	controllerNamespace := flag.String("controller-namespace", "linkerd", "namespace in which Linkerd is installed")
+	clusterDomain := flag.String("cluster-domain", "cluster.local", "custom cluster domain")
```

Comment:
This shouldn't be necessary, since the cluster domain is already persisted in the `linkerd-config` config map, which is mounted to the controller pods. 

Using a snippet from https://github.com/linkerd/linkerd2/pull/2930, in `controller/api/public/top_routes.go`, you can do something like:

```diff
diff --git a/controller/api/public/top_routes.go b/controller/api/public/top_routes.go
index 4195ce37..9e173b7a 100644
--- a/controller/api/public/top_routes.go
+++ b/controller/api/public/top_routes.go
@@ -10,6 +10,7 @@ import (
 	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
 	pb "github.com/linkerd/linkerd2/controller/gen/public"
 	api "github.com/linkerd/linkerd2/controller/k8s"
+	"github.com/linkerd/linkerd2/pkg/config"
 	"github.com/linkerd/linkerd2/pkg/k8s"
 	"github.com/prometheus/common/model"
 	log "github.com/sirupsen/logrus"
@@ -139,7 +140,13 @@ func (s *grpcServer) topRoutesFor(ctx context.Context, req *pb.TopRoutesRequest,
 		if err != nil {
 			return nil, err
 		}
+
 		// Find service profiles for all services in all objects in the resource.
+		global, err := config.Global(s.mountPathGlobalConfig)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving global config - %s", err)
+		}
+		clusterDomain := global.GetClusterDomain()
 		for _, obj := range objects {
 			// Lookup services for each object.
 			services, err := s.k8sAPI.GetServicesFor(obj, false)
@@ -148,7 +155,7 @@ func (s *grpcServer) topRoutesFor(ctx context.Context, req *pb.TopRoutesRequest,
 			}
 
 			for _, svc := range services {
-				p := s.k8sAPI.GetServiceProfileFor(svc, clientNs)
+				p := s.k8sAPI.GetServiceProfileFor(svc, clientNs, clusterDomain)
 				profiles[svc.GetName()] = p
 			}
 		}
```
As a matter of fact, we don't want users to be able to change the cluster domain to something different than that  in the `linkerd-config` config map.

Reply:
Very good suggestion, thanks!

It'll be a bit of work to integrate it into the split PRs though i'll see if i can make it work.

Reply:
Removed from this PR. Will be readded in the next.